exercises/Samuel_Jones/ex01_basics/src/main.cpp

```diff
@@ -0,0 +1,204 @@
+
+//Define namespace variables
+using namespace std;
+
+bool doesFileExist(string &);
+string getFileName();
+void readWords(vector<string> *, ifstream &);
+void findOccurencesofWords(vector<string>, vector<pair<string, int>> *);
+bool notDelim(char);
+bool sortingWords(const pair<string, int> &, const pair<string, int> &);
+
+/**
+ * Skeleton main routine
+ */
+
+int main(int, char **)
+{
+    //Get the file name from the user via command line
+    string fileName = getFileName();
+
+    //Load the ascii file
+    vector<string> fileStrings;
+    vector<pair<string, int>> uniqueWords;
+
+    ifstream fs;
+    fs.open(fileName.c_str());
+    //Read the file words and get each unique word in
+    readWords(&fileStrings, fs);
+
+    fs.close();
+
+    //Make sure that the unique words vector is in the correct order
+    findOccurencesofWords(fileStrings, &uniqueWords);
+
+    //Sort the lists based on the number of times it is present
+    sort(uniqueWords.begin(), uniqueWords.end(), sortingWords);
+
+    //print out to file named results, could request filename using getFileName() but not required.
+    ofstream results;
+    results.open("results.txt");
+    results << "Word\t\tUsage" << endl;
+    results << endl;
+    for (auto i = 0u; i < uniqueWords.size(); i++)
+    {
+        pair<string, int> currentWord = uniqueWords.at(i);
+        results << currentWord.first << "\t\t" << currentWord.second << endl;
+    }
+}
+
+//Get the file name a)nd check whether or not the file exists etc.
+string getFileName()
+{
+    string fileName;
+    do
+    {
+        fileName = "";
+        cout << "Give the  file name" << endl;
+        cin >> fileName;
+    } while (!doesFileExist(fileName));
+    return fileName;
+}
+
+//Return true if the file exists else false
+bool doesFileExist(string &fileName)
+{
+    std::ifstream infile(fileName.c_str());
+    if (!infile.good())
+    {
+        cout << "File could not be found!" << endl;
+    }
+    return infile.good();
+}
+
+void readWords(vector<string> *words, ifstream &fs)
+{
+    //Create the stream buffer for content into a string.
+    string contents((istreambuf_iterator<char>(fs)), istreambuf_iterator<char>());
+
+    //Create the word utilising the buffer iterators
+    string tempWord;
+    for (auto i = 0u; i < contents.length(); i++)
+    {
+        if (notDelim(contents[i]))
+        {
+            tempWord += contents[i];
+        }
+        else if (tempWord != "" && tempWord.length() >= 4)
+        {
+            //Make sure word is all lower case
+            transform(tempWord.begin(), tempWord.end(), tempWord.begin(), ::tolower);
+
+            words->push_back(tempWord);
+            tempWord = "";
+        }
+        else
+        {
+            tempWord = "";
+        }
+    }
+    //Catch the final word and add it to the list as it will be forgotten.
+    words->push_back(tempWord);
+}
+
+//If the charecter is one of the delimeters return false
+bool notDelim(char c)
+{
```

Comment:
The alternative to this is the following (where `c` is a single charterer):
` if(c >= 'a'&& c <= 'z')`
This will work and save you typing out the list below however you would have to be a bit careful of `\n` or similar.

Reply:
You make a fair point on this but I don't see a reason to change to get a little bit of performance out it when I would then just have to do further checks for the the end characters and null terminators.

Reply:
Performance is not the only metric by which we measure software quality. Especially in the majority of mantid. 
```

Reply:
Performance is not the only metric by which we measure software quality, especially in the majority of mantid.
```
static auto const punctuation = std::string(".,?'\\\"!():\n\r");
std::find(punctuation.cbegin(), punctuation.cend(), c) == punctuation.cend());
```
Is considerably less code, more readable and more maintainable than the original case statement.